ClientRuntimes/Java/azure-client-runtime/src/main/java/com/microsoft/azure/PagedList.java

```diff
@@ -0,0 +1,291 @@
+            }
+        }
+    }
+
+    /**
+     * The implementation of {@link Iterator} for PagedList.
+     */
+    private class Itr implements Iterator<E> {
+        /** The iterator for the actual list of items. */
+        private Iterator<E> itemsItr;
+
+        /**
+         * Creates an instance of the iterator.
+         */
+        public Itr() {
+            itemsItr = items.iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return itemsItr.hasNext() || nextPageLink != null;
+        }
+
+        @Override
+        public E next() {
+            if (!itemsItr.hasNext()) {
+                if (nextPageLink == null) {
+                    throw new NoSuchElementException();
+                } else {
+                    try {
+                        Page<E> nextPage = loadPage(nextPageLink);
+                        nextPageLink = nextPage.getNextPageLink();
+                        addAll(nextPage.getItems());
+                        itemsItr = items.iterator();
+                    } catch (CloudException e) {
+                        throw new WebServiceException(e.toString(), e);
+                    } catch (IOException e) {
+                        throw new DataBindingException(e.getMessage(), e);
+                    }
+                }
+            }
+            return itemsItr.next();
+        }
+
+        @Override
+        public void remove() {
+            itemsItr.remove();
+        }
+    }
+
+    /**
+     * The implementation of {@link ListIterator} for PagedList.
+     */
+    private class ListItr extends Itr implements ListIterator<E> {
+        /** The list iterator for the actual list of items. */
+        private ListIterator<E> itemsListItr;
+
+        /**
+         * Creates an instance of the ListIterator.
+         *
+         * @param index the position in the list to start.
+         */
+        public ListItr(int index) {
+            itemsListItr = items.listIterator(index);
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            return itemsListItr.hasPrevious();
+        }
+
+        @Override
+        public E previous() {
+            return itemsListItr.previous();
+        }
+
+        @Override
+        public int nextIndex() {
+            return itemsListItr.nextIndex();
+        }
+
+        @Override
+        public int previousIndex() {
+            return itemsListItr.previousIndex();
+        }
+
+        @Override
+        public void set(E e) {
+            itemsListItr.set(e);
+        }
+
+        @Override
+        public void add(E e) {
+            itemsListItr.add(e);
+        }
+    }
+
+    @Override
+    public int size() {
+        loadAll();
+        return items.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return items.isEmpty() && nextPageLink == null;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        loadAll();
+        return items.contains(o);
```

Comment:
Do we need to invoke loadAll() here. We could load page by page - check for the item in each page and return as soon as we find it.